### PR TITLE
[WIP] EventPipe reliability fix

### PR DESCRIPTION
**Pull request by cshung**

### CODE REVIEW

File: src/vm/eventpipebuffer.h [hunk 1]
```diff
@@ -110,8 +149,14 @@ class EventPipeBuffer
     // Get the next event from the buffer, but don't mark it read.
```
Comment (cshung) [inline src/vm/eventpipebuffer.h]: The "Don't mark it read" of the comment is now confusing. We expect the buffer to be in read-only mode, we are just not logically removing the event from the buffer.

File: src/vm/eventpipebuffermanager.cpp [hunk 2]
```diff
@@ -493,51 +561,156 @@ EventPipeEventInstance* EventPipeBufferManager::GetNextEvent()
...
+    {
+        THROWS;
+        GC_NOTRIGGER;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
 
+    _ASSERTE(EnsureConsistency());
+
+    // All calls to this method must not be synchronized by our caller
```
Comment (noahfalk) [inline src/vm/eventpipebuffermanager.cpp]:
```suggestion
    // All calls to this method must be synchronized by our caller
```

File: src/vm/eventpipebuffermanager.cpp [hunk 3]
```diff
@@ -732,54 +911,70 @@ unsigned int EventPipeBufferList::GetCount() const
...
+        if (this->m_pHeadBuffer->PeekNext(beforeTimeStamp))
         {
-            pNext = m_pReadBuffer->PeekNext(beforeTimeStamp);
+            // Case 3
+            return this->m_pHeadBuffer;
+        }
+        else
+        {
+            // Case 4
+            return this->m_pHeadBuffer->GetNext();
```
Comment (noahfalk) [inline src/vm/eventpipebuffermanager.cpp]: m_pHeadBuffer was older than beforeTimeStamp, but m_pHeadBuffer->GetNext() might not be.

File: src/vm/eventpipebuffermanager.h [hunk 4]
```diff
@@ -100,6 +122,24 @@ class EventPipeBufferManager
...
     bool WriteEvent(Thread *pThread, EventPipeSession &session, EventPipeEvent &event, EventPipeEventPayload &payload, LPCGUID pActivityId, LPCGUID pRelatedActivityId, Thread *pEventThread = NULL, StackContents *pStack = NULL);
 
+    // Ends the suspension period created by SuspendWriteEvent(). After this call returns WriteEvent()
+    // can again be called succesfully, new BufferLists and Buffers may be allocated.
+    // The caller is required to synchronize all calls to SuspendWriteEvent() and ResumeWriteEvent()
+    void ResumeWriteEvent();
+
+    // From the time this function returns until ResumeWriteEvent() is called a suspended state will 
+    // be in effect that blocks all WriteEvent activity. All existing buffers will be in the 
+    // PENDING_FLUSH state and no new EventPipeBuffers or EventPipeBufferLists can be created. Calls to
```
Comment (noahfalk) [inline src/vm/eventpipebuffermanager.h]:
```suggestion
    // READ_ONLY state and no new EventPipeBuffers or EventPipeBufferLists can be created. Calls to 
```

A relic of some earlier naming I had

File: src/vm/eventpipe.cpp [hunk 5]
```diff
@@ -741,15 +752,11 @@ void EventPipe::WriteEventInternal(EventPipeEvent &event, EventPipeEventPayload
     }
     else if (s_pConfig->RundownEnabled())
     {
-        // It is possible that some events that are enabled on rundown can be emitted from other threads.
-        // We're not interested in these events and they can cause corrupted trace files because rundown
-        // events are written synchronously and not under lock.
-        // If we encounter an event that did not originate on the thread that is doing rundown, ignore it.
-        if (pThread == NULL || !s_pConfig->IsRundownThread(pThread))
+        if (pThread == nullptr)
```
[the remark this answers is not in the document]
Comment (jorive) [reply src/vm/eventpipe.cpp]: Have we verified this?

**Comment by noahfalk** [comment]
This PR fixes https://github.com/dotnet/coreclr/issues/22502.

**Pull request merged**